RecoVertex/PrimaryVertexProducer/plugins/PrimaryVertexProducer.cc

```diff
@@ -41,53 +42,84 @@ PrimaryVertexProducer::PrimaryVertexProducer(const edm::ParameterSet& conf)
+  } else if (clusteringAlgorithm == "DA_vect") {
     theTrackClusterizer = new DAClusterizerInZ_vect(
         conf.getParameter<edm::ParameterSet>("TkClusParameters").getParameter<edm::ParameterSet>("TkDAClusParameters"));
   } else if (clusteringAlgorithm == "DA2D_vect") {
     theTrackClusterizer = new DAClusterizerInZT_vect(
         conf.getParameter<edm::ParameterSet>("TkClusParameters").getParameter<edm::ParameterSet>("TkDAClusParameters"));
-    f4D = true;
-  }
-
-  else {
+    useTransientTrackTime = true;
+  } else {
     throw VertexException("PrimaryVertexProducer: unknown clustering algorithm: " + clusteringAlgorithm);
   }
 
-  if (f4D) {
-    trkTimesToken = consumes<edm::ValueMap<float>>(conf.getParameter<edm::InputTag>("TrackTimesLabel"));
-    trkTimeResosToken = consumes<edm::ValueMap<float>>(conf.getParameter<edm::InputTag>("TrackTimeResosLabel"));
+  if (useTransientTrackTime) {
+    trkTimesToken = consumes<edm::ValueMap<float> >(conf.getParameter<edm::InputTag>("TrackTimesLabel"));
+    trkTimeResosToken = consumes<edm::ValueMap<float> >(conf.getParameter<edm::InputTag>("TrackTimeResosLabel"));
+    trackMTDTimeQualityToken =
+        consumes<edm::ValueMap<float> >(conf.getParameter<edm::InputTag>("trackMTDTimeQualityVMapTag"));
+    minTrackTimeQuality = conf.getParameter<double>("minTrackTimeQuality");
   }
 
   // select and configure the vertex fitters
   std::vector<edm::ParameterSet> vertexCollections =
-      conf.getParameter<std::vector<edm::ParameterSet>>("vertexCollections");
+      conf.getParameter<std::vector<edm::ParameterSet> >("vertexCollections");
 
   for (std::vector<edm::ParameterSet>::const_iterator algoconf = vertexCollections.begin();
        algoconf != vertexCollections.end();
        algoconf++) {
     algo algorithm;
+
+    algorithm.label = algoconf->getParameter<std::string>("label");
+
+    // configure the fitter and selector
     std::string fitterAlgorithm = algoconf->getParameter<std::string>("algorithm");
     if (fitterAlgorithm == "KalmanVertexFitter") {
-      algorithm.fitter = new KalmanVertexFitter();
+      algorithm.pv_fitter = new SequentialPrimaryVertexFitterAdapter(new KalmanVertexFitter());
     } else if (fitterAlgorithm == "AdaptiveVertexFitter") {
-      algorithm.fitter = new AdaptiveVertexFitter(GeometricAnnealing(algoconf->getParameter<double>("chi2cutoff")));
+      auto fitter = new AdaptiveVertexFitter(GeometricAnnealing(algoconf->getParameter<double>("chi2cutoff")));
+      algorithm.pv_fitter = new SequentialPrimaryVertexFitterAdapter(fitter);
+    } else if (fitterAlgorithm.empty()) {
+      algorithm.pv_fitter = nullptr;
+    } else if (fitterAlgorithm == "AdaptiveChisquareVertexFitter") {
+      algorithm.pv_fitter = new AdaptiveChisquarePrimaryVertexFitter(algoconf->getParameter<double>("chi2cutoff"),
+                                                                     algoconf->getParameter<double>("zcutoff"),
```

Comment:
@werdmann @kdeleo I see this new parameter added, and propagated to ```AdaptiveChisquarePrimaryVertexFitter```, but I fail to see where it is effectively used. Is this intended for future use? Or a leftover of developments?

Reply:
It is only used in the multivertex mode. The single vertex mode is constructed such that it reproduces the  AdaptiveVertexFitter as much as possible.

Reply:
@fabiocos So, yes,  in this sense it is for future use and not currently part of any baseline workflow.
The parameter is uses as z_cutoff_ here:
https://github.com/kdeleo/cmssw/blob/d546464ab24166f93f81e4370b929cfae20cda94/RecoVertex/PrimaryVertexProducer/src/AdaptiveChisquarePrimaryVertexFitter.cc#L127

Reply:
@werdmann ok, thanks, I had missed it